### DISCUSSION TITLE
Issue with Building Eclipse 2023-12 Based Product using Tycho 4.0.4

### DISCUSSION THREAD

**Discussion by cristianvalentincojocaru** [eclipse-tycho/tycho#4228] [Q&A]
Hello,

We are currently attempting to build an **Eclipse 2023-12**-based product but have encountered issues with the build process. Despite numerous efforts to resolve the problem, we consistently encounter the same error. What's strange is that we were able to successfully build a similar product using **Eclipse 2023-06 with no issues**. The only significant change between the two builds is the version of Eclipse p2 repository used for dependency resolution.

**Setup Details**:
- **Tycho version**: 4.0.4
- **Maven version**: 3.9.6
- **Java version**: 17.0.9

**Error Log:**

```
 The actions required to successfully install the requested software can not be installed. 
  Cannot complete the install because of a conflicting dependency.
  	Software being installed: org.eclipse.equinox.p2.engine.actions.root.DefaultProfile 1.0.0.1725884057854
  	Software currently installed: org.eclipse.tycho.p2 2.7.5 (tycho-bundles-external 2.7.5)
  	Only one of the following can be installed at once: 
  		Common Eclipse Runtime 3.18.200.v20231106-1826 (org.eclipse.equinox.common 3.18.200.v20231106-1826)
  		Common Eclipse Runtime 3.16.0.v20220211-2322 (org.eclipse.equinox.common 3.16.0.v20220211-2322)
  	Cannot satisfy dependency:
  		From: S32 Design Studio extensions for Eclipse P2 Core 3.6.0.202409091142 (com.ext.rcp.p2.core 3.6.0.202409091142)
  		To: osgi.bundle; org.eclipse.core.filesystem 0.0.0
  	Cannot satisfy dependency:
  		From: Core File Systems 1.10.200.v20231102-0934 (org.eclipse.core.filesystem 1.10.200.v20231102-0934)
  		To: osgi.bundle; org.eclipse.equinox.common [3.18.0,4.0.0)
  	Cannot satisfy dependency:
  		From: org.eclipse.equinox.p2.engine.actions.root.DefaultProfile 1.0.0.1725884057854
  		To: org.eclipse.equinox.p2.iu; com.ext.rcp.p2.core 1.0.0
  	Cannot satisfy dependency:
  		From: org.eclipse.tycho.p2 2.7.5 (tycho-bundles-external 2.7.5)
  		To: org.eclipse.equinox.p2.iu; org.eclipse.equinox.common [3.16.0.v20220211-2322,3.16.0.v20220211-2322]
```

**Observations**:

- **Build Success with 2023-06**: The same product was successfully built using the Eclipse 2023-06 p2 repository (https://download.eclipse.org/eclipse/updates/4.28/) for dependency resolution without any issues. For the newest build we tried using the following p2 repository for Eclipse 2023-12: https://download.eclipse.org/releases/2023-12/
- **Conflicting Dependencies:** The error seems to orbit around conflicting versions of org.eclipse.equinox.common, where both 3.18.200 and 3.16.0 are being pulled in, and Tycho is unable to resolve the conflict.
- We found a similar issue ([Issue #1325](https://github.com/eclipse-tycho/tycho/issues/1325)) and tried **incrementing the Tycho version** (4.0.5 and 4.0.8), but it did not resolve the problem.
- Another issue related to **custom p2 touchpoint actions** ([Issue #1250](https://github.com/eclipse-tycho/tycho/issues/1250)) resulted in the same error, but we believe it may not be relevant to our case, as we already built a fully functional product with the 2023-06 repository, without any change of the already implemented actions

**Request for Assistance:**
We suspect this issue may be related to a compatibility problem between the Eclipse 2023-12 p2 repository and Tycho. However, we would greatly appreciate any insights or recommendations on how to address this conflict. 

Best regards,  
Cristian

**Comment by laeubi** [comment]
It looks like you somehow pull in `tycho-bundles-external` what no longer seem to function in your specific case (but it is hard to guess without an example).

As custom touchpoints are currently not really well supported there is a good chance this is the issue you are facing here, so I would check if custom touchpoints are crucial for your use case. If they are important for you, you might want to debugging / improving custom-touchpoints or try to reach out to [Tycho professional support options](https://github.com/eclipse-tycho/tycho?tab=readme-ov-file#getting-support).

Beside that, using the latest Tycho 4 release should be the very first step in any case, even if it not help it makes sure you are as close to a possible bugfix as possible already.

**Reply by cristianvalentincojocaru** [reply]
Thank you for the suggestion. Do you have any additional recommendations besides removing the custom touchpoints? We would prefer to consider that option as a last resort, as we have quite a few of them, and their removal would be quite disruptive.

In this case, we were able to build a fully functional product using the Eclipse 2023-06 p2 repository without making any changes to the custom touchpoints. The build setup, including Tycho, Maven, and Java versions, is identical for both builds. The only difference is the switch to the Eclipse 2023-12 p2 repository, where we encountered the issue.

Given that the setup worked with 2023-06, we’re hoping there may be an alternative solution that avoids changing the touchpoints. Any further guidance would be appreciated!

**Reply by cristianvalentincojocaru** [reply]
After further investigation, we were able to successfully create a product build locally by manually running the following command, but **with one change**. We used the **org.eclipse.equinox.launcher_1.6.600.v20231106-1826.jar** (from Eclipse 2023-12) instead of the **org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar** that the original command was using. Other than this change to the Equinox launcher version, everything else in the setup remained the same.

Here’s the original command that was using the older version of the launcher:
`/opt/Tools/jre-17.0.9/bin/java -jar ${pathToProdTargeDirPlugin}/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar -application org.eclipse.equinox.p2.director -metadataRepository ${pathTo}/repository/ -artifactRepository ${pathTo}/repository/ -installIU ${listOfFeaturesToInstall} -destination ${destination} -profile DefaultProfile -profileProperties org.eclipse.update.install.features=true -roaming -p2.os linux -p2.ws gtk -p2.arch x86_64`

Do you have any idea why this specific version of the Equinox launcher is used during the installation phase? What could be different between the versions, and why would one work while the other does not?

Any insights would be greatly appreciated.
